pyls_mypy/plugin.py

```diff
@@ -41,17 +44,45 @@ def parse_line(line, document=None):
 
 @hookimpl
 def pyls_lint(config, document):
+    args = ['--incremental',
+            '--show-column-numbers',
+            '--follow-imports', 'silent']
+
+    config_files = config.plugin_settings('pyls_mypy').get('config_files', [])
+    config_files = (os.path.expandvars(os.path.expanduser(p.strip()))
+                    for p in config_files)
+    config_files = [p for p in config_files if p]
+    while config_files:
+        # This seems a little complicated but the idea is that configuration
+        # files in relative paths that are not separated by absolute path
+        # configuration files, will all have an equal chance of being found
+        # in each directory before the search moves up the project tree.
+        config_file = None
+        relative = list(itertools.takewhile(lambda p: not os.path.isabs(p),
+                                            config_files))
+        if relative:
+            config_files = config_files[len(relative):]
+
+            found = config.find_parents(os.getcwd(), relative)
+            if not found:
+                found = [p for p in relative if os.path.exists(p)]
+
+            if found:
+                config_file = found[0]
+        else:
+            value = config_files.pop(0)
+            if os.path.exists(value):
+                config_file = value
+
+        if config_file is not None:
+            args += ['--config-file', config_file]
+            break
```

Comment:
I am a bit confused by this algorithm of choosing a file. IIUC, you simply want to find the first configuration file that exists in the system. There's no need to modify `config_files` variable at all. You simply iterate with `for one_file in config_files:`, and then for the `one_file` you do the checks: whether it's relative, and whether it exists.

For bonus points, if you move the whole check to a separate function, you would eliminate the need to ever mutate `config_file` *(not the `config_files` ;))* variable. It will be as simple as: if file exists, return it; if not, return `None`.

Reply:
No, that would result in a relative pathed config file higher up the project directory tree in taking precedence over one that's lower, if it appears first in the config file list. This is not ideal as you want all the relative config files highest in the directory tree (i.e. closest to the point of invocation) to be considered first before any further down the tree.

This algorithm ensures that each group of relative pathed files appearing before every group of absolute pathed files will take equal precedence as we scan up the project tree.

Reply:
Oh, I see